cmd/forwarder/config.go

```diff
@@ -79,3 +80,17 @@ func (c IssConfig) LogAuthUser(user string, pct int) bool {
 		user != c.ValidTokenUser &&
 		pct <= c.TokenUserSamplePct
 }
+
+func (c *IssConfig) TokenMap() map[string]string {
+	if c.tokenMap == nil {
```

Comment:
Consider using [sync.Once](https://golang.org/pkg/sync/#Once) to make this safe to be called by multiple goroutines in the nil case.

Reply:
@danp This pushes the limit of my minimal go experience here. When first looking at the `sync.Once` example in the godoc, I thought maybe the following would suffice.
```go
func (c *IssConfig) TokenMap() map[string]string {
	var once sync.Once
	once.Do(func() {
		if c.tokenMap == nil {
			pairs := strings.Split(c.Tokens, "|")
			c.tokenMap = make(map[string]string)

			for _, pair := range pairs {
				unameToken := strings.Split(pair, ":")
				c.tokenMap[unameToken[0]] = unameToken[1]
			}
		}
	}())

	return c.tokenMap
}
```

But it seems to be that given that `once` is defined on every call, it will simply just create multiple instances of it and defeat the purpose, maybe?

Do you mind guiding me a bit more here? Should I make `once` a global, setting it side of the function, or perhaps set it in the `IssConfig` struct? Or will the above work?

Reply:
Adding something like `tokenMapOnce sync.Once` to the IssConfig struct and using `c.tokenMapOnce.Do(...)` should do the trick.

Reply:
This seems to be okay...
```go
var tokenMapOnce sync.Once

func (c *IssConfig) TokenMap() map[string]string {
	tmExtract := func() {
		if c.tokenMap == nil {
			pairs := strings.Split(c.Tokens, "|")
			c.tokenMap = make(map[string]string)

			for _, pair := range pairs {
				unameToken := strings.Split(pair, ":")
				c.tokenMap[unameToken[0]] = unameToken[1]
			}
		}
	}
	tokenMapOnce.Do(tmExtract)

	return c.tokenMap
}
```

Reply:
I would recommend doing it on the struct so it's not influenced by global state. Using a package variable like that there would be confusion if there were ever two IssConfigs. That's not likely to happen in production but could make testing more difficult.

Reply:
@danp 

It appears that adding it to `IssConfig` breaks `env.Decode`
```
$ DEPLOY=local FORWARD_DEST=localhost:1514 PORT=3000 TOKEN_MAP="foo:foo|bar:bar" ./forwarder
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method

goroutine 1 [running]:
reflect.valueInterface(0x132aba0, 0xc42010a0f0, 0x36, 0x1089d01, 0x134c700, 0x132aba0)
        /usr/local/go/src/reflect/value.go:959 +0x1c1
reflect.Value.Interface(0x132aba0, 0xc42010a0f0, 0x36, 0x132aba0, 0xc42010a0f0)
        /usr/local/go/src/reflect/value.go:948 +0x44
github.com/heroku/log-iss/vendor/github.com/joeshaw/envdecode.Decode(0x1331840, 0xc42010a000, 0x12a14ed, 0xc42009825c)
        /Users/jmervine/.go/src/github.com/heroku/log-iss/vendor/github.com/joeshaw/envdecode/envdecode.go:75 +0xed3
main.NewIssConfig(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jmervine/.go/src/github.com/heroku/log-iss/cmd/forwarder/config.go:41 +0x7a
main.main()
        /Users/jmervine/.go/src/github.com/heroku/log-iss/cmd/forwarder/main.go:32 +0xbf
```

It's the only change, when I move it to a package var, it works fine. I agree that it's better to have it in the struct for future proofing, but to avoid shaving yaks, I'm going to make it a package variable for now, unless you have strong objections.

Again, thanks for your assistance on this.

Reply:
Can you bump envdecode? I fixed that exact issue in https://github.com/joeshaw/envdecode/pull/20 :)

Reply:
Here's the diff that generated that error: https://gist.github.com/jmervine/1e56f81530a41060d0fb7400b30ef17e.